File: taurus-protect-sdk-typescript/src/services/job-service.ts
/**
 * Job service for Taurus-PROTECT SDK.
 *
 * Provides methods for monitoring jobs in the Taurus-PROTECT system.
 * Jobs are background tasks that process various operations such as
 * transaction monitoring, balance updates, and other async operations.
 */

import { NotFoundError, ValidationError } from '../errors';
import type { JobsApi } from '../internal/openapi/apis/JobsApi';
import { jobFromDto, jobsFromDto, jobStatusFromDto } from '../mappers/job';
import type { Job, JobStatus } from '../models/job';
import { BaseService } from './base';

/**
 * Service for monitoring jobs in the Taurus-PROTECT system.
 *
 * Jobs are background tasks that process various operations such as
 * transaction monitoring, balance updates, and other async operations.
 *
 * @example
 * ```typescript
 * // Get all jobs
 * const jobs = await jobService.list();
 * for (const job of jobs) {
 *   console.log(`${job.name}: ${job.statistics?.successes} successes`);
 * }
 *
 * // Get a specific job
 * const job = await jobService.get('balance-sync');
 * console.log(`Job ${job.name} has ${job.statistics?.pending} pending`);
 *
 * // Get job execution status
 * const status = await jobService.getStatus('balance-sync', 'exec-123');
 * console.log(`Status: ${status.status}, Message: ${status.message}`);
 * ```
 */
export class JobService extends BaseService {
  private readonly jobsApi: JobsApi;

  /**
   * Creates a new JobService instance.
   *
   * @param jobsApi - The JobsApi instance from the OpenAPI client
   */
  constructor(jobsApi: JobsApi) {
    super();
    this.jobsApi = jobsApi;
  }

  /**
   * Lists all jobs.
   *
   * @returns Array of jobs
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const jobs = await jobService.list();
   * for (const job of jobs) {
   *   console.log(`${job.name}: ${job.statistics?.successes} successes`);
   * }
   * ```
   */
  async list(): Promise<Job[]> {
    return this.execute(async () => {
      const response = await this.jobsApi.jobServiceGetJobs();

      const result =
        (response as Record<string, unknown>).jobs ??
        (response as Record<string, unknown>).result;
      return jobsFromDto(result as unknown[]);
    });
  }

  /**
   * Gets a job by name.
   *
   * @param name - The job name
   * @returns The job
   * @throws {@link ValidationError} If name is empty
   * @throws {@link NotFoundError} If job is not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const job = await jobService.get('balance-sync');
   * console.log(`Job ${job.name} has ${job.statistics?.pending} pending`);
   * ```
   */
  async get(name: string): Promise<Job> {
    if (!name || name.trim() === '') {
      throw new ValidationError('name is required');
    }

    return this.execute(async () => {
      const response = await this.jobsApi.jobServiceGetJob({ name });

      const result =
        (response as Record<string, unknown>).job ??
        (response as Record<string, unknown>).result;
      const job = jobFromDto(result);

      if (!job) {
        throw new NotFoundError(`Job with name '${name}' not found`);
      }

      return job;
    });
  }

  /**
   * Gets the status of a specific job execution.
   *
   * @param name - The job name
   * @param id - The job execution ID
   * @returns The job status
   * @throws {@link ValidationError} If name or id is empty
   * @throws {@link NotFoundError} If job status is not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const status = await jobService.getStatus('balance-sync', 'exec-123');
   * console.log(`Status: ${status.status}`);
   * console.log(`Started: ${status.startedAt}`);
   * console.log(`Message: ${status.message}`);
   * ```
   */
  async getStatus(name: string, id: string): Promise<JobStatus> {
    if (!name || name.trim() === '') {
      throw new ValidationError('name is required');
    }
    if (!id || id.trim() === '') {
      throw new ValidationError('id is required');
    }

    return this.execute(async () => {
      const response = await this.jobsApi.jobServiceGetJobStatus({ name, id });

      const result =
        (response as Record<string, unknown>).status ??
        (response as Record<string, unknown>).result;
      const status = jobStatusFromDto(result);

      if (!status) {
        throw new NotFoundError(`Job status for '${name}' execution '${id}' not found`);
      }

      return status;
    });
  }
}

File: taurus-protect-sdk-java/client/src/main/java/com/taurushq/sdk/protect/client/service/JobService.java
package com.taurushq.sdk.protect.client.service;

import com.taurushq.sdk.protect.client.mapper.ApiExceptionMapper;
import com.taurushq.sdk.protect.client.mapper.JobMapper;
import com.taurushq.sdk.protect.client.model.ApiException;
import com.taurushq.sdk.protect.client.model.Job;
import com.taurushq.sdk.protect.client.model.JobStatus;
import com.taurushq.sdk.protect.openapi.ApiClient;
import com.taurushq.sdk.protect.openapi.api.JobsApi;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetJobReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetJobStatusReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetJobsReply;

import java.util.List;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Service for monitoring jobs in the Taurus Protect system.
 * <p>
 * Jobs are background tasks that process various operations such as
 * transaction monitoring, balance updates, and other async operations.
 * <p>
 * Example usage:
 * <pre>{@code
 * // Get all jobs
 * List<Job> jobs = client.getJobService().getJobs();
 *
 * // Get a specific job
 * Job job = client.getJobService().getJob("balance-sync");
 *
 * // Get job execution status
 * JobStatus status = client.getJobService().getJobStatus("balance-sync", "exec-123");
 * }</pre>
 *
 * @see Job
 * @see JobStatus
 */
public class JobService {

    private final JobsApi jobsApi;
    private final ApiExceptionMapper apiExceptionMapper;
    private final JobMapper jobMapper;

    /**
     * Creates a new JobService.
     *
     * @param apiClient          the API client for making HTTP requests
     * @param apiExceptionMapper the mapper for converting API exceptions
     * @throws NullPointerException if any parameter is null
     */
    public JobService(final ApiClient apiClient, final ApiExceptionMapper apiExceptionMapper) {
        checkNotNull(apiClient, "apiClient must not be null");
        checkNotNull(apiExceptionMapper, "apiExceptionMapper must not be null");
        this.jobsApi = new JobsApi(apiClient);
        this.apiExceptionMapper = apiExceptionMapper;
        this.jobMapper = JobMapper.INSTANCE;
    }

    /**
     * Retrieves all jobs.
     *
     * @return the list of jobs
     * @throws ApiException if the API call fails
     */
    public List<Job> getJobs() throws ApiException {
        try {
            TgvalidatordGetJobsReply reply = jobsApi.jobServiceGetJobs();
            return jobMapper.fromDTOList(reply.getJobs());
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves a job by name.
     *
     * @param name the job name
     * @return the job
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if name is null or empty
     */
    public Job getJob(final String name) throws ApiException {
        checkArgument(name != null && !name.isEmpty(), "name must not be null or empty");
        try {
            TgvalidatordGetJobReply reply = jobsApi.jobServiceGetJob(name);
            return jobMapper.fromDTO(reply.getJob());
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves the status of a specific job execution.
     *
     * @param name the job name
     * @param id   the job execution ID
     * @return the job status
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if name or id is null or empty
     */
    public JobStatus getJobStatus(final String name, final String id) throws ApiException {
        checkArgument(name != null && !name.isEmpty(), "name must not be null or empty");
        checkArgument(id != null && !id.isEmpty(), "id must not be null or empty");
        try {
            TgvalidatordGetJobStatusReply reply = jobsApi.jobServiceGetJobStatus(name, id);
            return jobMapper.fromStatusDTO(reply.getStatus());
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }
}
